algorithms/kernel/svm/oneapi/svm_train_thunder_oneapi_impl.i

```diff
@@ -0,0 +1,277 @@
+}
+
+template <typename algorithmFPType, typename ParameterType>
+services::Status SVMTrainOneAPI<algorithmFPType, ParameterType, thunder>::smoKernel(
+    const services::Buffer<algorithmFPType> & y, const services::Buffer<algorithmFPType> & kernelWsRows, const services::Buffer<uint32_t> & wsIndices,
+    const uint32_t ldK, const services::Buffer<algorithmFPType> & f, const algorithmFPType C, const algorithmFPType eps, const algorithmFPType tau,
+    const uint32_t maxInnerIteration, services::Buffer<algorithmFPType> & alpha, services::Buffer<algorithmFPType> & deltaalpha,
+    services::Buffer<algorithmFPType> & resinfo, const size_t nWS)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(smoKernel);
+
+    auto & context = services::Environment::getInstance()->getDefaultExecutionContext();
+    auto & factory = context.getClKernelFactory();
+
+    services::String build_options = getKeyFPType<algorithmFPType>();
+
+    services::String cachekey("__daal_algorithms_svm_smo_block_");
+    build_options.add(" -D WS_SIZE=");
+    char bufferString[DAAL_MAX_STRING_SIZE] = { 0 };
+    services::daal_int_to_string(bufferString, DAAL_MAX_STRING_SIZE, int(nWS));
+    build_options.add(bufferString);
+    build_options.add(" -D SIMD_WIDTH=64 ");
+    cachekey.add(build_options);
+
+    services::Status status;
+    factory.build(ExecutionTargetIds::device, cachekey.c_str(), clKernelBlockSMO, build_options.c_str(), &status);
+
+    DAAL_CHECK_STATUS_VAR(status);
+
+    auto kernel = factory.getKernel("smoKernel");
+
+    KernelArguments args(12);
+    args.set(0, y, AccessModeIds::read);
+    args.set(1, kernelWsRows, AccessModeIds::read);
+    args.set(2, wsIndices, AccessModeIds::read);
+    args.set(3, ldK);
+    args.set(4, f, AccessModeIds::read);
+    args.set(5, C);
+    args.set(6, eps);
+    args.set(7, tau);
+    args.set(8, maxInnerIteration);
+    args.set(9, alpha, AccessModeIds::readwrite);
+    args.set(10, deltaalpha, AccessModeIds::readwrite);
+    args.set(11, resinfo, AccessModeIds::readwrite);
+
+    KernelRange localRange(nWS);
+    KernelRange globalRange(nWS);
+
+    KernelNDRange range(1);
+    range.global(globalRange, &status);
+    DAAL_CHECK_STATUS_VAR(status);
+    range.local(localRange, &status);
+    DAAL_CHECK_STATUS_VAR(status);
+
+    context.run(range, kernel, args, &status);
+    DAAL_CHECK_STATUS_VAR(status);
+
+    return status;
+}
+
+template <typename algorithmFPType, typename ParameterType>
+bool SVMTrainOneAPI<algorithmFPType, ParameterType, thunder>::checkStopCondition(const algorithmFPType diff, const algorithmFPType diffPrev,
+                                                                                 const algorithmFPType eps, const size_t nNoChanges,
+                                                                                 size_t & sameLocalDiff)
+{
+    sameLocalDiff = utils::internal::abs(diff - diffPrev) < eps * 1e-2 ? sameLocalDiff + 1 : 0;
```

Comment:
What is rationale behind multiplying `eps` by `0.01`?

Reply:
We use two exit criteria:
1. When the gradient difference in the subtask is less than epsilon.
2. We must control that several external iterations should not lead to the same result - this is achieved by this criterion. But so that there is no quick exit, we deliberately reduce epsilon for this criteria.